File: content/tracks/dsa-typescript/concepts/11-binary-search-trees/binary-search-trees_test.ts
import { TreeNode, isValidBST, kthSmallest, buildBSTFromPreorder } from "./stub";

let passed = 0;
let failed = 0;

function test(name: string, fn: () => void): void {
  try {
    fn();
    console.log(`  PASS: ${name}`);
    passed++;
  } catch (e: unknown) {
    const msg = e instanceof Error ? e.message : String(e);
    console.log(`  FAIL: ${name} — ${msg}`);
    failed++;
  }
}

function assertEqual<T>(actual: T, expected: T): void {
  if (actual !== expected) {
    throw new Error(`expected ${JSON.stringify(expected)}, got ${JSON.stringify(actual)}`);
  }
}

// Inorder traversal helper for verifying buildBSTFromPreorder
function inorder(root: TreeNode | null): number[] {
  if (root === null) return [];
  return [...inorder(root.left), root.val, ...inorder(root.right)];
}

// --- isValidBST ---

test("isValidBST: valid BST", () => {
  const bst = new TreeNode(2, new TreeNode(1), new TreeNode(3));
  assertEqual(isValidBST(bst), true);
});

test("isValidBST: right child smaller than root", () => {
  const t = new TreeNode(5, new TreeNode(1), new TreeNode(4, new TreeNode(3), new TreeNode(6)));
  assertEqual(isValidBST(t), false);
});

test("isValidBST: null tree", () => {
  assertEqual(isValidBST(null), true);
});

test("isValidBST: single node", () => {
  assertEqual(isValidBST(new TreeNode(1)), true);
});

test("isValidBST: node violates ancestor bound (classic trap)", () => {
  // root=10, left=5, left.right=15 — 15 is in left subtree but > root 10
  const t = new TreeNode(10, new TreeNode(5, null, new TreeNode(15)), new TreeNode(20));
  assertEqual(isValidBST(t), false);
});

test("isValidBST: larger valid BST", () => {
  const bst = new TreeNode(5,
    new TreeNode(3, new TreeNode(2), new TreeNode(4)),
    new TreeNode(6)
  );
  assertEqual(isValidBST(bst), true);
});

test("isValidBST: equal values (not a strict BST)", () => {
  const t = new TreeNode(2, new TreeNode(2), null);
  assertEqual(isValidBST(t), false);
});

// --- kthSmallest ---

test("kthSmallest: k=1 is minimum", () => {
  const bst = new TreeNode(3, new TreeNode(1, null, new TreeNode(2)), new TreeNode(4));
  assertEqual(kthSmallest(bst, 1), 1);
});

test("kthSmallest: k=2", () => {
  const bst = new TreeNode(3, new TreeNode(1, null, new TreeNode(2)), new TreeNode(4));
  assertEqual(kthSmallest(bst, 2), 2);
});

test("kthSmallest: k=3 (root)", () => {
  const bst = new TreeNode(3, new TreeNode(1, null, new TreeNode(2)), new TreeNode(4));
  assertEqual(kthSmallest(bst, 3), 3);
});

test("kthSmallest: k=n (maximum)", () => {
  const bst = new TreeNode(5, new TreeNode(3), new TreeNode(6));
  assertEqual(kthSmallest(bst, 3), 6);
});

test("kthSmallest: single node", () => {
  assertEqual(kthSmallest(new TreeNode(42), 1), 42);
});

// --- buildBSTFromPreorder ---

test("buildBSTFromPreorder: inorder of result is sorted", () => {
  const preorder = [8, 5, 1, 7, 10, 12];
  const root = buildBSTFromPreorder(preorder);
  // Inorder of a valid BST must be sorted
  const result = inorder(root);
  const expected = [...preorder].sort((a, b) => a - b);
  assertEqual(JSON.stringify(result), JSON.stringify(expected));
});

test("buildBSTFromPreorder: root is preorder[0]", () => {
  const root = buildBSTFromPreorder([8, 5, 1, 7, 10, 12]);
  assertEqual(root?.val ?? null, 8);
});

test("buildBSTFromPreorder: single element", () => {
  const root = buildBSTFromPreorder([5]);
  assertEqual(root?.val ?? null, 5);
  assertEqual(root?.left ?? null, null);
  assertEqual(root?.right ?? null, null);
});

test("buildBSTFromPreorder: result is a valid BST", () => {
  const root = buildBSTFromPreorder([8, 5, 1, 7, 10, 12]);
  assertEqual(isValidBST(root), true);
});

test("buildBSTFromPreorder: right-skewed preorder", () => {
  // Ascending preorder → right-skewed BST
  const root = buildBSTFromPreorder([1, 2, 3]);
  assertEqual(inorder(root).join(","), "1,2,3");
});

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);

File: content/tracks/dsa-python/concepts/11-binary-search-trees/binary-search-trees_test.py
import sys
from stub import TreeNode, is_valid_bst, kth_smallest, build_bst_from_preorder, _build

passed = 0
failed = 0


def test(name: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        print(f"  PASS: {name}")
        passed += 1
    else:
        msg = f" — {detail}" if detail else ""
        print(f"  FAIL: {name}{msg}")
        failed += 1


def inorder(root: "TreeNode | None") -> list[int]:
    result: list[int] = []
    def dfs(node: "TreeNode | None") -> None:
        if node is None:
            return
        dfs(node.left)
        result.append(node.val)
        dfs(node.right)
    dfs(root)
    return result


# is_valid_bst
test("valid_bst: simple valid [2,1,3]", is_valid_bst(_build([2, 1, 3])) is True)
test("valid_bst: right child invalid", is_valid_bst(_build([5, 1, 4, None, None, 3, 6])) is False)
test("valid_bst: empty tree", is_valid_bst(None) is True)
test("valid_bst: single node", is_valid_bst(_build([1])) is True)
test("valid_bst: cross-subtree trap [10,5,15,null,null,6,20]",
     is_valid_bst(_build([10, 5, 15, None, None, 6, 20])) is False)
test("valid_bst: deep valid", is_valid_bst(_build([5, 4, 6, 3, None, None, 7])) is True)

# kth_smallest
test("kth: k=1 in [3,1,4,null,2]", kth_smallest(_build([3, 1, 4, None, 2]), 1) == 1)
test("kth: k=3 in [3,1,4,null,2]", kth_smallest(_build([3, 1, 4, None, 2]), 3) == 3)
test("kth: k=3 in [5,3,6,2,4,null,null,1]", kth_smallest(_build([5, 3, 6, 2, 4, None, None, 1]), 3) == 3)
test("kth: single node k=1", kth_smallest(_build([1]), 1) == 1)
test("kth: k=4 in [3,1,4,null,2]", kth_smallest(_build([3, 1, 4, None, 2]), 4) == 4)

# build_bst_from_preorder
r = build_bst_from_preorder([8, 5, 1, 7, 10, 12])
test("build_preorder: root is 8", r is not None and r.val == 8, f"got {r.val if r else None}")
test("build_preorder: inorder is sorted", inorder(r) == [1, 5, 7, 8, 10, 12], f"got {inorder(r)}")
r2 = build_bst_from_preorder([1])
test("build_preorder: single element", r2 is not None and r2.val == 1 and r2.left is None and r2.right is None)
r3 = build_bst_from_preorder([])
test("build_preorder: empty list", r3 is None)
r4 = build_bst_from_preorder([4, 2, 1, 3, 6, 5, 7])
test("build_preorder: balanced BST inorder", inorder(r4) == [1, 2, 3, 4, 5, 6, 7], f"got {inorder(r4)}")

print(f"\n{passed} passed, {failed} failed")
if failed > 0:
    sys.exit(1)
